### PR TITLE
bug/image processing issues

### PR DESCRIPTION
**Pull request by Gnzlt**
## What does this implement/fix? Explain your changes.
Fix [image processing issues](https://github.com/meganz/android2/pull/1454#issuecomment-705948434) reported by @Piasy 

## Any relevant logs, error output, etc?
...

## Any other comments?
...

## Where has this been tested?

- **Devices/Simulators:** 
Google Pixel 4

- **OS Version:** 
Android 11

### CODE REVIEW

**Comment by javiergm1983** [comment]
@Piasy could you check if your [reported issues here](https://github.com/meganz/android2/pull/1454#issuecomment-705948434) are finally solved with this PR?
If so, please review the code.

**Comment by Piasy** [comment]
After a quick test, seems only this one is fixed:
> when take 2 pictures, rotate the second picture 4 times, which makes it to the original orientation, then crop it, the picture isn't shown in crop screen

this one has a new issue:

> when take a picture, rotate it, then crop it, the preview isn't rotated in crop screen

I can't adjust the left/right edge of the crop area.

Please check it when you have time @Gnzlt .

**Comment by Gnzlt** [comment]
> in crop screen, user can set a non-rectangle crop area, but the result image is rectangle, which will stretch the content, is it designed behavior?
The cropping implies a perspective correction based on OpenCV so it will make adjustments every time you decide a new area, it's expected.
> the picture isn't displayed in center when discard one picture, please check this screen recording,
This should not happen anymore
> it always scroll from the first picture to the last one after taking new picture, please check this screen recording,

**Comment by Gnzlt** [comment]
> in crop screen, user can set a non-rectangle crop area, but the result image is rectangle, which will stretch the content, is it designed behavior?

The cropping implies a perspective correction based on OpenCV so it will make adjustments every time you decide a new area, it's expected.
> the picture isn't displayed in center when discard one picture, please check this screen recording,

This should not happen anymore
> it always scroll from the first picture to the last one after taking new picture, please check this screen recording,

This is the expected behavior
> when take a picture, rotate it, then crop it, the preview isn't rotated in crop screen

This should not happen anymore

> I can't adjust the left/right edge of the crop area.
Working on it

**Comment by javiergm1983** [comment]
@Gnzlt told me the changes are ready. @Piasy please could you finish your review? Thanks.

**Comment by Piasy** [comment]
Hi I test with `com.github.meganz:AndroidDocumentScanner:35f38e2`, and this issue is still there, please check them when you have time.

> the picture isn't displayed in center when discard one picture, please check this screen recording, https://mega.nz/file/H94GBCxZ#uRjfEmrvWy074ETaWl8Sj5oX-KOcFS5px2n0qclUvdU

All other issues are resolved, except https://jira.developers.mega.co.nz/browse/AND-12535, which shouldn't block this PR now.

Code changes look good, but I just find out it still use Glide, could you please replace it with Fresco? I don't agree with Javier on this issue:

> I suppose that @Gnzlt uses Glide because is the one originally used in the library (maybe also because could be better for this scenario). As this is a separate project from the app, in principle I don't see a problem to use Glide instead Fresco here, the only one thing we have to take into account is use only one library for image processing in the project.

At first, the original library doesn't use Glide, secondly, if we use Glide at this project, then Glide will also be included into our final MEGA app, which will increase APK size.

But if you plan to replace Glide with Fresco in another separate PR, I think it's fine.

**Comment by javiergm1983** [comment]
You are right @Piasy , we should try to use Fresco here as well like in whole app project to use only one image procesing library and don't increase the APK size much more.

**Comment by Gnzlt** [comment]
As mentioned on the original review, my first plan was to use Fresco but I ran into some issues that I needed to spend more time on. Do you guys think I should tackle those issues now? Or instead doing the migration afterwards to unblock this feature?
> Initially I planned to use Fresco, and I actually migrated the calls to Fresco because of that same reason, trying to avoid having multiple libraries for the same purpose. But while I was getting into more complex tasks where it required bitmap rotation or grayscale conversion, I ended up having some OOMs. At that point I decided to migrate back to Glide since I didn't have those issues. In any case this could be change in the future since I didn't spend that much time trying to solve those issues with Fresco.

**Comment by javiergm1983** [comment]
Please make a new try to see if you achieve migrate to Fresco avoiding those issues you mentioned. It is also important don't increase the APK size much more than strictly needed.

You can create a new branch from this for that try, so if it is finally unsuccesfull we could easy discard it.

**Comment by Gnzlt** [comment]
Fresco migration has been merged and page offset has been fixed. This is ready for review @Piasy

**Comment by Piasy** [comment]
> the picture isn't displayed in center when discard one picture, please check this screen recording, https://mega.nz/file/H94GBCxZ#uRjfEmrvWy074ETaWl8Sj5oX-KOcFS5px2n0qclUvdU

Hi @Gnzlt , this issue still happens on my Huawei Honor V30 Android 10 and Nexus 5X Android 8.1, does it work as expected on your device? BTW, I test it inside MEGA application, do you test it in the demo app or this repo?

**Comment by Gnzlt** [comment]
> I test it inside MEGA application, do you test it in the demo app or this repo?
This should be fixed, and works on my Moto G5S too. If you test it inside MEGA, keep in mind that Gradle might be building it with a cached version of the repo. I would recommend you to test it on this repo. Once this is merged, I will merge develop changes and request a review on MEGA PR.

**Review by Piasy — DISMISSED** [review]
test demo app in the separate repo, the issue doesn't happen.

File: documentscanner/src/main/java/nz/mega/documentscanner/DocumentScannerViewModel.kt
Comment (javiergm1983) [inline documentscanner/src/main/java/nz/mega/documentscanner/DocumentScannerViewModel.kt:117]: Add method documentation.

File: documentscanner/src/main/java/nz/mega/documentscanner/DocumentScannerViewModel.kt
Comment (javiergm1983) [inline documentscanner/src/main/java/nz/mega/documentscanner/DocumentScannerViewModel.kt:154]: Add method documentation.

File: documentscanner/src/main/java/nz/mega/documentscanner/DocumentScannerViewModel.kt
Comment (javiergm1983) [inline documentscanner/src/main/java/nz/mega/documentscanner/DocumentScannerViewModel.kt:182]: Add method documentation.

File: documentscanner/src/main/java/nz/mega/documentscanner/DocumentScannerViewModel.kt
Comment (javiergm1983) [inline documentscanner/src/main/java/nz/mega/documentscanner/DocumentScannerViewModel.kt:207]: Add method documentation.

File: documentscanner/src/main/java/nz/mega/documentscanner/DocumentScannerViewModel.kt
Comment (javiergm1983) [inline documentscanner/src/main/java/nz/mega/documentscanner/DocumentScannerViewModel.kt:214]: Add method documentation.

File: documentscanner/src/main/java/nz/mega/documentscanner/camera/CameraFragment.kt
Comment (javiergm1983) [inline documentscanner/src/main/java/nz/mega/documentscanner/camera/CameraFragment.kt:183]: Add method documentation.

File: documentscanner/src/main/java/nz/mega/documentscanner/openCV/CropDetector.kt
Comment (javiergm1983) [inline documentscanner/src/main/java/nz/mega/documentscanner/openCV/CropDetector.kt:25]: Add method documentation.

File: documentscanner/src/main/java/nz/mega/documentscanner/openCV/CropDetector.kt
Comment (javiergm1983) [inline documentscanner/src/main/java/nz/mega/documentscanner/openCV/CropDetector.kt:40]: Add method documentation.

File: documentscanner/src/main/java/nz/mega/documentscanner/openCV/CropDetector.kt
Comment (javiergm1983) [inline documentscanner/src/main/java/nz/mega/documentscanner/openCV/CropDetector.kt:113]: Add method documentation.

File: documentscanner/src/main/java/nz/mega/documentscanner/openCV/CropDetector.kt
Comment (javiergm1983) [inline documentscanner/src/main/java/nz/mega/documentscanner/openCV/CropDetector.kt:140]: Add method documentation.

File: documentscanner/src/main/java/nz/mega/documentscanner/openCV/ImageScanner.kt
Comment (javiergm1983) [inline documentscanner/src/main/java/nz/mega/documentscanner/openCV/ImageScanner.kt:22]: Add method documentation.

File: documentscanner/src/main/java/nz/mega/documentscanner/openCV/ImageScanner.kt
Comment (javiergm1983) [inline documentscanner/src/main/java/nz/mega/documentscanner/openCV/ImageScanner.kt:28]: Add method documentation.

File: documentscanner/src/main/java/nz/mega/documentscanner/openCV/ImageScanner.kt
Comment (javiergm1983) [inline documentscanner/src/main/java/nz/mega/documentscanner/openCV/ImageScanner.kt:37]: Add method documentation.

File: documentscanner/src/main/java/nz/mega/documentscanner/openCV/OpenCvUtils.kt
Comment (javiergm1983) [inline documentscanner/src/main/java/nz/mega/documentscanner/openCV/OpenCvUtils.kt:52]: Add method documentation.

File: documentscanner/src/main/java/nz/mega/documentscanner/scan/ScanFragment.kt
Comment (javiergm1983) [inline documentscanner/src/main/java/nz/mega/documentscanner/scan/ScanFragment.kt:118]: Add method documentation.

File: documentscanner/src/main/java/nz/mega/documentscanner/scan/ScanFragment.kt
Comment (javiergm1983) [inline documentscanner/src/main/java/nz/mega/documentscanner/scan/ScanFragment.kt:150]: Add method documentation.

File: documentscanner/src/main/java/nz/mega/documentscanner/scan/ScanFragment.kt
Comment (javiergm1983) [inline documentscanner/src/main/java/nz/mega/documentscanner/scan/ScanFragment.kt:157]: Add method documentation.

File: documentscanner/src/main/java/nz/mega/documentscanner/utils/BitmapUtils.kt
Comment (javiergm1983) [inline documentscanner/src/main/java/nz/mega/documentscanner/utils/BitmapUtils.kt:33]: Add method documentation.

File: documentscanner/src/main/java/nz/mega/documentscanner/utils/BitmapUtils.kt
Comment (javiergm1983) [inline documentscanner/src/main/java/nz/mega/documentscanner/utils/BitmapUtils.kt:56]: Add method documentation.

File: documentscanner/src/main/java/nz/mega/documentscanner/utils/BitmapUtils.kt
Comment (javiergm1983) [inline documentscanner/src/main/java/nz/mega/documentscanner/utils/BitmapUtils.kt:63]: Add method documentation.

File: documentscanner/src/main/java/nz/mega/documentscanner/utils/BitmapUtils.kt
Comment (javiergm1983) [inline documentscanner/src/main/java/nz/mega/documentscanner/utils/BitmapUtils.kt:73]: Add method documentation.

File: documentscanner/src/main/java/nz/mega/documentscanner/utils/BitmapUtils.kt
Comment (javiergm1983) [inline documentscanner/src/main/java/nz/mega/documentscanner/utils/BitmapUtils.kt:85]: Add method documentation.

File: documentscanner/src/main/java/nz/mega/documentscanner/utils/BitmapUtils.kt
Comment (javiergm1983) [inline documentscanner/src/main/java/nz/mega/documentscanner/utils/BitmapUtils.kt:91]: Add method documentation.

**Review by javiergm1983 — COMMENTED** [review]
Hi @Gnzlt, as Piasy said, the issues seems fixed and everything  seem work fine. However there are several new and modified methods with no documentation.

As we have commented several times, although it is based on a 3rd party library is important have a well documented code to help us in future fixes or developments, so is important add documentation to the new methods included or the modified ones.

I have pointed some methods where I miss documentation, please take a look and add it.

File: documentscanner/src/main/java/nz/mega/documentscanner/openCV/CropDetector.kt
Comment (Gnzlt) [reply documentscanner/src/main/java/nz/mega/documentscanner/openCV/CropDetector.kt:40]: Methods on this class are related with the OpenCV algorithm inherited from the forked lib. It was created by the original author who separated the logic based on his own criteria.

File: documentscanner/src/main/java/nz/mega/documentscanner/openCV/CropDetector.kt
Comment (Gnzlt) [reply documentscanner/src/main/java/nz/mega/documentscanner/openCV/CropDetector.kt:113]: Methods on this class are related with the OpenCV algorithm inherited from the forked lib. It was created by the original author who separated the logic based on his own criteria.

File: documentscanner/src/main/java/nz/mega/documentscanner/openCV/CropDetector.kt
Comment (Gnzlt) [reply documentscanner/src/main/java/nz/mega/documentscanner/openCV/CropDetector.kt:140]: Methods on this class are related with the OpenCV algorithm inherited from the forked lib. It was created by the original author who separated the logic based on his own criteria.

File: documentscanner/src/main/java/nz/mega/documentscanner/openCV/CropDetector.kt
Comment (javiergm1983) [reply documentscanner/src/main/java/nz/mega/documentscanner/openCV/CropDetector.kt:47]: Could we have at least a minimum documentation explaining that and the parameter meaning?

File: documentscanner/src/main/java/nz/mega/documentscanner/openCV/CropDetector.kt
Comment (javiergm1983) [reply documentscanner/src/main/java/nz/mega/documentscanner/openCV/CropDetector.kt:120]: Could we have at least a minimum documentation explaining that and the parameters meaning?

File: documentscanner/src/main/java/nz/mega/documentscanner/openCV/CropDetector.kt
Comment (javiergm1983) [reply documentscanner/src/main/java/nz/mega/documentscanner/openCV/CropDetector.kt:147]: Could we have at least a minimum documentation explaining that?

**Review by javiergm1983 — APPROVED** [review]

**Comment by javiergm1983** [comment]
@Piasy please approve again and merge this PR.

**Review by Piasy — APPROVED** [review]
Everything looks fine now, approved.

**Pull request merged**
